### ISSUE TITLE
Pyright does not seem to equate `is False` with `== False` with union return types

### ISSUE THREAD

**Issue by NCPlayz** [microsoft/pyright#2008]
**Describe the bug**
When describing the return type as `Literal[False] | ...`, it seems that checking if the value `is False` is not equated to `== False`, causing access errors later on. 

**To Reproduce**
```py
from typing import Literal, Union

def x() -> Union[str, Literal[False]]:
    raise NotImplementedError

def test():
    value = x()

    if value is False:
        return
    
    reveal_type(value)

    value.capitalize()
```

**Expected behavior**
type of `value` should be `str`.

**Screenshots or Code**

![image](https://i.imgur.com/1rOjuDk.png)
![image](https://i.imgur.com/xVDL9EO.png)

**VS Code extension or command-line**
`ms-python.vscode-pylance` `v2021.6.2`

**Comment by erictraut** [comment]
Thanks for the feature request. You're correct, pyright doesn't currently handle type narrowing for this pattern. Here is a [list of the patterns](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards) that are supported.

Each new pattern that we support for conditional type narrowing adds complexity and potentially slows down analysis, so we're selective in which patterns we support. In particular, we need some evidence that it's a pattern that appears frequently in Python code. There are two signals we use to inform this decision: 1) we receive input from a significant number of users that they want support for the pattern, or 2) other Python type checkers have already added support this pattern.

In this particular case, I confirmed that mypy already supports this particular pattern, so it meets the bar.

**Comment by erictraut** [comment]
This is implemented and will be included in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.151, which I just published. It will also be included in the next release of pylance.
